diff --git a/src/h/impl/math_fenv.h b/src/h/impl/math_fenv.h
index 0a93c8e0a..9630f4c6c 100644
--- a/src/h/impl/math_fenv.h
+++ b/src/h/impl/math_fenv.h
@@ -72,15 +72,14 @@
 
 #if defined(HAVE_FENV_H) && !defined(ECL_AVOID_FPE_H)
 # if defined(HAVE_FEENABLEEXCEPT)
-#  define ECL_WITH_LISP_FPE_BEGIN do {                   \
-        fenv_t __fenv;                                   \
-        fegetenv(&__fenv);                               \
-        cl_env_ptr __the_env = ecl_process_env_unsafe(); \
-        if (__the_env) {                                 \
-                int bits = __the_env->trap_fpe_bits;     \
-                fedisableexcept(FE_ALL_EXCEPT & ~bits);  \
-                feenableexcept(FE_ALL_EXCEPT & bits);    \
-        }                                                \
+#  define ECL_WITH_LISP_FPE_BEGIN do {                       \
+        fenv_t __fenv;                                       \
+        fegetenv(&__fenv);                                   \
+        if (ecl_get_option(ECL_OPT_BOOTED) > 0) {            \
+                int bits = ecl_process_env()->trap_fpe_bits; \
+                fedisableexcept(FE_ALL_EXCEPT & ~bits);      \
+                feenableexcept(FE_ALL_EXCEPT & bits);        \
+        }                                                    \
         feclearexcept(FE_ALL_EXCEPT);
 # else
 #  define ECL_WITH_LISP_FPE_BEGIN do {                   \
