diff --git a/src/sage/all.py b/src/sage/all.py
index 3ef717df659..26f1902f8ea 100644
--- a/src/sage/all.py
+++ b/src/sage/all.py
@@ -187,6 +187,12 @@ from sage.misc.all import *
 from sage.doctest.all import *
 from sage.repl.all import *
 
+
+# This import also sets up the interrupt handler
+from cysignals.signals import (AlarmInterrupt, SignalError,
+                               sig_on_reset as sig_on_count)
+
+from time import sleep
 from functools import reduce  # in order to keep reduce in python3
 
 import sage.misc.lazy_import
@@ -213,6 +219,7 @@ from sage.sat.all import *
 from sage.schemes.all import *
 from sage.graphs.all import *
 from sage.groups.all import *
+from sage.arith.power import generic_power as power
 from sage.databases.all import *
 from sage.categories.all import *
 from sage.sets.all import *
@@ -279,6 +286,8 @@ if sys.platform != 'win32':
 # Lazily import interacts (#15335)
 lazy_import('sage.interacts', 'all', 'interacts')
 
+from copy import copy, deepcopy
+
 # The code executed here uses a large amount of Sage components
 from sage.rings.qqbar import _init_qqbar
 _init_qqbar()
diff --git a/src/sage/all__sagemath_environment.py b/src/sage/all__sagemath_environment.py
new file mode 100644
index 00000000000..4baf97bae83
--- /dev/null
+++ b/src/sage/all__sagemath_environment.py
@@ -0,0 +1,4 @@
+# sage_setup: distribution = sagemath-environment
+from sage.env import SAGE_ROOT, SAGE_SRC, SAGE_DOC_SRC, SAGE_LOCAL, DOT_SAGE, SAGE_ENV
+
+from sage.misc.all__sagemath_environment import *
diff --git a/src/sage/all__sagemath_objects.py b/src/sage/all__sagemath_objects.py
new file mode 100644
index 00000000000..17d32f635c1
--- /dev/null
+++ b/src/sage/all__sagemath_objects.py
@@ -0,0 +1,33 @@
+# sage_setup: distribution = sagemath-objects
+import os
+import sys
+import operator
+import math
+
+import warnings
+
+# TODO: More to be moved from all.py
+
+# This import also sets up the interrupt handler
+from cysignals.signals import (AlarmInterrupt, SignalError,
+                               sig_on_reset as sig_on_count)
+
+from time import sleep
+
+from sage.misc.all__sagemath_objects import *
+from sage.structure.all import *
+from sage.arith.power import generic_power as power
+from sage.categories.all__sagemath_objects import *
+
+from sage.cpython.all import *
+
+from copy import copy, deepcopy
+
+true = True
+false = False
+
+
+# For doctesting. These are overwritten later
+
+Integer = int
+RealNumber = float
diff --git a/src/sage/all__sagemath_repl.py b/src/sage/all__sagemath_repl.py
new file mode 100644
index 00000000000..093dc376749
--- /dev/null
+++ b/src/sage/all__sagemath_repl.py
@@ -0,0 +1,118 @@
+# sage_setup: distribution = sagemath-repl
+
+# Set up warning filters before importing Sage stuff
+
+import sys
+import warnings
+
+# This is a Python debug build (--with-pydebug)
+__with_pydebug = hasattr(sys, 'gettotalrefcount')
+if __with_pydebug:
+    # a debug build does not install the default warning filters. Sadly, this breaks doctests so we
+    # have to re-add them:
+    warnings.filterwarnings('ignore', category=PendingDeprecationWarning)
+    warnings.filterwarnings('ignore', category=ImportWarning)
+    warnings.filterwarnings('ignore', category=ResourceWarning)
+else:
+    deprecationWarning = ('ignore', None, DeprecationWarning, None, 0)
+    if deprecationWarning in warnings.filters:
+        warnings.filters.remove(deprecationWarning)
+
+# Ignore all deprecations from IPython etc.
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        module='(IPython|ipykernel|jupyter_client|jupyter_core|nbformat|notebook|ipywidgets|storemagic|jedi)')
+
+# scipy 1.18 introduced deprecation warnings on a number of things they are moving to
+# numpy, e.g. DeprecationWarning: scipy.array is deprecated
+#             and will be removed in SciPy 2.0.0, use numpy.array instead
+# This affects networkx 2.2 up and including 2.4 (cf. :issue:29766)
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        module='(scipy|networkx)')
+
+# However, be sure to keep OUR deprecation warnings
+warnings.filterwarnings('default', category=DeprecationWarning,
+                        message=r'[\s\S]*See https?://trac\.sagemath\.org/[0-9]* for details.')
+
+# Ignore Python 3.9 deprecation warnings
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        module='ast')
+
+# Ignore packaging 20.5 deprecation warnings
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        module='(.*[.]_vendor[.])?packaging')
+
+# Ignore a few warnings triggered by pythran 0.12.1
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message='\n\n  `numpy.distutils` is deprecated since NumPy 1.23.0',
+                        module='pythran.dist')
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message='pkg_resources is deprecated as an API|'
+                        'Deprecated call to `pkg_resources.declare_namespace(.*)`',
+                        module='pkg_resources|setuptools.sandbox')
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message='msvccompiler is deprecated and slated to be removed',
+                        module='distutils.msvccompiler')
+
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message='The distutils(.sysconfig module| package) is deprecated',
+                        module='Cython|distutils|numpy|sage.env|sage.features')
+
+# triggered by cython 0.29.32
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message="'cgi' is deprecated and slated for removal in Python 3.13",
+                        module='Cython')
+
+# triggered by pyparsing 2.4.7
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message="module 'sre_constants' is deprecated",
+                        module='pyparsing')
+
+# importlib.resources.path and ...read_binary are deprecated in python 3.11,
+# but the replacement importlib.resources.files needs python 3.9
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message=r'(path|read_binary) is deprecated\. Use files\(\) instead\.',
+                        module='sage.repl.rich_output.output_(graphics|graphics3d|video)')
+
+# triggered by sphinx
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message="'imghdr' is deprecated and slated for removal in Python 3.13",
+                        module='sphinx.util.images')
+
+# triggered by docutils 0.19 on Python 3.11
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message=r"Use setlocale\(\), getencoding\(\) and getlocale\(\) instead",
+                        module='docutils.io')
+
+# triggered by dateutil 2.8.2 and sphinx 7.0.1 on Python 3.12
+# see: https://github.com/dateutil/dateutil/pull/1285
+# see: https://github.com/sphinx-doc/sphinx/pull/11468
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message=r"datetime.datetime.utcfromtimestamp\(\) is deprecated",
+                        module='dateutil.tz.tz|sphinx.(builders.gettext|util.i18n)')
+
+# triggered on Python 3.12
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message=r"This process.* is multi-threaded, "
+                                r"use of .*\(\) may lead to deadlocks in the child.")
+
+# pickling of itertools is deprecated in Python 3.12
+warnings.filterwarnings('ignore', category=DeprecationWarning,
+                        message=r"Pickle, copy, and deepcopy support will be "
+                                r"removed from itertools in Python 3.14.")
+
+# rpy2>=3.6 emits warnings for R modifying LD_LIBRARY_PATH
+warnings.filterwarnings('ignore', category=UserWarning,
+                        message=r".*redefined by R and overriding existing variable.*",
+                        module='rpy2.*')
+
+from sage.all__sagemath_objects import *
+from sage.all__sagemath_environment import *
+
+from sage.doctest.all import *
+from sage.repl.all import *
+from sage.misc.all__sagemath_repl import *
+
+# For doctesting. These are overwritten later
+
+Integer = int
+RealNumber = float
diff --git a/src/sage/all__sagemath_tdlib.py b/src/sage/all__sagemath_tdlib.py
new file mode 100644
index 00000000000..e26612c8409
--- /dev/null
+++ b/src/sage/all__sagemath_tdlib.py
@@ -0,0 +1 @@
+# sage_setup: distribution = sagemath-tdlib
diff --git a/src/sage/arith/all__sagemath_objects.py b/src/sage/arith/all__sagemath_objects.py
new file mode 100644
index 00000000000..40da49fdd73
--- /dev/null
+++ b/src/sage/arith/all__sagemath_objects.py
@@ -0,0 +1,3 @@
+# sage_setup: distribution = sagemath-objects
+# The presence of this file ensures that sage_setup for sagemath-objects
+# considers this directory as a namespace package
diff --git a/src/sage/arith/meson.build b/src/sage/arith/meson.build
index f0e9395524a..500b49edf85 100644
--- a/src/sage/arith/meson.build
+++ b/src/sage/arith/meson.build
@@ -1,6 +1,7 @@
 py.install_sources(
   '__init__.py',
   'all.py',
+  'all__sagemath_objects.py',
   'constants.pxd',
   'functions.pxd',
   'long.pxd',
diff --git a/src/sage/categories/all.py b/src/sage/categories/all.py
index 53f7ba085d1..7393a5ee028 100644
--- a/src/sage/categories/all.py
+++ b/src/sage/categories/all.py
@@ -32,33 +32,7 @@ from sage.categories import primer
 
 from sage.misc.lazy_import import lazy_import
 
-# Resolve a circular import so that "import sage.categories.all" can succeed
-# in initializing the category system.
-import sage.structure.category_object  # imports sage.categories.category
-
-# Small part of "from sage.categories.basic import *":
-from sage.categories.objects import Objects
-from sage.categories.sets_cat import Sets, EmptySetError
-
-
-from sage.categories.category import Category
-
-from sage.categories.category_types import Elements
-
-from sage.categories.cartesian_product import cartesian_product
-
-from sage.categories.functor import (ForgetfulFunctor,
-                                     IdentityFunctor)
-
-from sage.categories.homset import (Hom, hom,
-                                    End, end,
-                                    Homset, HomsetWithBase)
-
-from sage.categories.morphism import Morphism
-
-from sage.categories.realizations import Realizations
-
-from sage.categories.sets_with_partial_maps import SetsWithPartialMaps
+from sage.categories.all__sagemath_objects import *
 
 from sage.categories.basic import *
 
diff --git a/src/sage/categories/all__sagemath_objects.py b/src/sage/categories/all__sagemath_objects.py
new file mode 100644
index 00000000000..42d003d22e1
--- /dev/null
+++ b/src/sage/categories/all__sagemath_objects.py
@@ -0,0 +1,33 @@
+# sage_setup: distribution = sagemath-objects
+# Subset of sage.categories.all that is made available by the sage-objects distribution
+
+from sage.misc.lazy_import import lazy_import
+
+# Resolve a circular import so that "import sage.categories.all" can succeed
+# in initializing the category system.
+import sage.structure.category_object  # imports sage.categories.category
+
+# Small part of "from sage.categories.basic import *":
+from sage.categories.objects import Objects
+from sage.categories.sets_cat import Sets, EmptySetError
+
+
+from sage.categories.category import Category
+
+from sage.categories.category_types import Elements
+
+from sage.categories.cartesian_product import cartesian_product
+
+from sage.categories.functor import (ForgetfulFunctor,
+                                     IdentityFunctor)
+
+from sage.categories.homset import (Hom, hom,
+                                    End, end,
+                                    Homset, HomsetWithBase)
+
+from sage.categories.morphism import Morphism
+
+from sage.categories.realizations import Realizations
+
+from sage.categories.sets_with_partial_maps import SetsWithPartialMaps
+del lazy_import
diff --git a/src/sage/categories/meson.build b/src/sage/categories/meson.build
index 43b9c696f8e..642cee8e9e7 100644
--- a/src/sage/categories/meson.build
+++ b/src/sage/categories/meson.build
@@ -12,6 +12,7 @@ py.install_sources(
   'algebras.py',
   'algebras_with_basis.py',
   'all.py',
+  'all__sagemath_objects.py',
   'aperiodic_semigroups.py',
   'associative_algebras.py',
   'basic.py',
diff --git a/src/sage/meson.build b/src/sage/meson.build
index 8e2ee496123..d1d15cf6895 100644
--- a/src/sage/meson.build
+++ b/src/sage/meson.build
@@ -82,6 +82,10 @@ endforeach
 py.install_sources(
   '__init__.py',
   'all.py',
+  'all__sagemath_environment.py',
+  'all__sagemath_objects.py',
+  'all__sagemath_repl.py',
+  'all__sagemath_tdlib.py',
   'all_cmdline.py',
   'env.py',
   'version.py',
diff --git a/src/sage/misc/all.py b/src/sage/misc/all.py
index 4102073d48d..c2d4ab35267 100644
--- a/src/sage/misc/all.py
+++ b/src/sage/misc/all.py
@@ -1,43 +1,9 @@
 from sage.misc.lazy_attribute import lazy_attribute, lazy_class_attribute
 from sage.misc.lazy_import import lazy_import
 
-import sage.structure.all   # to break a cyclic import
-
-from sage.misc.verbose import (set_verbose, set_verbose_files,
-                               get_verbose_files, unset_verbose_files, get_verbose)
-lazy_import('sage.misc.verbose', 'verbose',
-            deprecation=17815)
-from sage.misc.call import attrcall
-
-from sage.misc.misc_c import prod, running_total, balanced_sum
-mul = prod
-add = sum
-
-from sage.misc.repr import repr_lincomb
-
-from sage.misc.flatten import flatten
-
-from sage.misc.persist import save, load, dumps, loads, db, db_save
-
-from sage.misc.constant_function import ConstantFunction
-
-from sage.misc.sage_unittest import TestSuite
-
-from sage.misc.decorators import specialize, sage_wraps, infix_operator
-
-from sage.misc.unknown import Unknown, UnknownError
-
-from sage.misc.cachefunc import CachedFunction, cached_function, cached_method, cached_in_parent_method, disk_cached_function
-
-from sage.misc.abstract_method import abstract_method
-
-from sage.misc.timing import walltime, cputime
-
-from sage.misc.temporary_file import tmp_dir, tmp_filename
-
-from sage.misc.sage_eval import sage_eval, sageobj
-
-from sage.misc.sage_input import sage_input
+from sage.misc.all__sagemath_objects import *
+from sage.misc.all__sagemath_environment import *
+from sage.misc.all__sagemath_repl import *
 
 from sage.misc.misc import (
     exists, forall, is_iterator, random_sublist, pad_zeros,
@@ -150,6 +116,8 @@ from sage.misc.randstate import seed, set_random_seed, initial_seed, current_ran
 
 from sage.misc.prandom import *
 
+from sage.misc.timing import walltime, cputime
+
 from sage.misc.explain_pickle import explain_pickle, unpickle_newobj, unpickle_global, unpickle_build, unpickle_instantiate, unpickle_persistent, unpickle_extension, unpickle_appends
 
 lazy_import('sage.misc.inline_fortran', 'fortran')
diff --git a/src/sage/misc/all.py.rej b/src/sage/misc/all.py.rej
new file mode 100644
index 00000000000..8db1badbabf
--- /dev/null
+++ b/src/sage/misc/all.py.rej
@@ -0,0 +1,52 @@
+--- src/sage/misc/all.py
++++ src/sage/misc/all.py
+@@ -1,46 +1,9 @@
+ from sage.misc.lazy_attribute import lazy_attribute, lazy_class_attribute
+ from sage.misc.lazy_import import lazy_import
+ 
+-import sage.structure.all   # to break a cyclic import
+-
+-from sage.misc.lazy_attribute import lazy_attribute, lazy_class_attribute
+-from sage.misc.lazy_import import lazy_import
+-
+-from sage.misc.verbose import (set_verbose, set_verbose_files,
+-                               get_verbose_files, unset_verbose_files, get_verbose)
+-lazy_import('sage.misc.verbose', 'verbose',
+-            deprecation=17815)
+-from sage.misc.call import attrcall
+-
+-from sage.misc.misc_c import prod, running_total, balanced_sum
+-mul = prod
+-add = sum
+-
+-from sage.misc.repr import repr_lincomb
+-
+-from sage.misc.flatten import flatten
+-
+-from sage.misc.persist import save, load, dumps, loads, db, db_save
+-
+-from sage.misc.constant_function import ConstantFunction
+-
+-from sage.misc.sage_unittest import TestSuite
+-
+-from sage.misc.decorators import specialize, sage_wraps, infix_operator
+-
+-from sage.misc.unknown import Unknown, UnknownError
+-
+-from sage.misc.cachefunc import CachedFunction, cached_function, cached_method, cached_in_parent_method, disk_cached_function
+-
+-from sage.misc.abstract_method import abstract_method
+-
+-from sage.misc.timing import walltime, cputime
+-
+-from sage.misc.temporary_file import tmp_dir, tmp_filename
+-
+-from sage.misc.sage_eval import sage_eval, sageobj
+-
+-from sage.misc.sage_input import sage_input
++from sage.misc.all__sagemath_objects import *
++from sage.misc.all__sagemath_environment import *
++from sage.misc.all__sagemath_repl import *
+ 
+ from sage.misc.misc import (BackslashOperator,
+                   exists, forall, is_iterator,
diff --git a/src/sage/misc/all__sagemath_environment.py b/src/sage/misc/all__sagemath_environment.py
new file mode 100644
index 00000000000..2f09e9c4d56
--- /dev/null
+++ b/src/sage/misc/all__sagemath_environment.py
@@ -0,0 +1,2 @@
+# sage_setup: distribution = sagemath-environment
+from sage.misc.temporary_file import tmp_dir, tmp_filename
diff --git a/src/sage/misc/all__sagemath_objects.py b/src/sage/misc/all__sagemath_objects.py
new file mode 100644
index 00000000000..634584c5567
--- /dev/null
+++ b/src/sage/misc/all__sagemath_objects.py
@@ -0,0 +1,37 @@
+# sage_setup: distribution = sagemath-objects
+# Subset of sage.misc.all that is made available by the sage-objects distribution
+
+import sage.structure.all   # to break a cyclic import
+
+from sage.misc.lazy_attribute import lazy_attribute, lazy_class_attribute
+from sage.misc.lazy_import import lazy_import
+
+from sage.misc.verbose import (set_verbose, set_verbose_files,
+                               get_verbose_files, unset_verbose_files, get_verbose)
+lazy_import('sage.misc.verbose', 'verbose',
+            deprecation=17815)
+from sage.misc.call import attrcall
+
+from sage.misc.misc_c import prod, running_total, balanced_sum
+mul = prod
+add = sum
+
+from sage.misc.repr import repr_lincomb
+
+from sage.misc.flatten import flatten
+
+from sage.misc.persist import save, load, dumps, loads, db, db_save
+
+from sage.misc.constant_function import ConstantFunction
+
+from sage.misc.sage_unittest import TestSuite
+
+from sage.misc.decorators import specialize, sage_wraps, infix_operator
+
+from sage.misc.unknown import Unknown, UnknownError
+
+from sage.misc.cachefunc import CachedFunction, cached_function, cached_method, cached_in_parent_method, disk_cached_function
+
+from sage.misc.abstract_method import abstract_method
+
+from sage.misc.timing import walltime, cputime
diff --git a/src/sage/misc/all__sagemath_repl.py b/src/sage/misc/all__sagemath_repl.py
new file mode 100644
index 00000000000..9631c49d0bd
--- /dev/null
+++ b/src/sage/misc/all__sagemath_repl.py
@@ -0,0 +1,4 @@
+# sage_setup: distribution = sagemath-repl
+from sage.misc.sage_eval import sage_eval, sageobj
+
+from sage.misc.sage_input import sage_input
diff --git a/src/sage/misc/lazy_import.pyx b/src/sage/misc/lazy_import.pyx
index 426fc563396..9bac24a371e 100644
--- a/src/sage/misc/lazy_import.pyx
+++ b/src/sage/misc/lazy_import.pyx
@@ -492,17 +492,17 @@ cdef class LazyImport():
 
         Here we show how to take a function in a module, and lazy
         import it as a method of a class. For the sake of this
-        example, we add manually a function in :mod:`sage.all`::
+        example, we add manually a function in :mod:`sage.all__sagemath_objects`::
 
             sage: def my_method(self): return self
-            sage: import sage.all
-            sage: sage.all.my_method = my_method
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.my_method = my_method
 
         Now we lazy import it as a method of a new class ``Foo``::
 
             sage: from sage.misc.lazy_import import LazyImport
             sage: class Foo():
-            ....:     my_method = LazyImport('sage.all', 'my_method')
+            ....:     my_method = LazyImport('sage.all__sagemath_objects', 'my_method')
 
         Now we can use it as a usual method::
 
@@ -579,13 +579,13 @@ cdef class LazyImport():
         TESTS::
 
             sage: from sage.misc.lazy_import import LazyImport
-            sage: import sage.all
-            sage: sage.all.foo = list(range(10))
-            sage: lazy_foo = LazyImport('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = list(range(10))
+            sage: lazy_foo = LazyImport('sage.all__sagemath_objects', 'foo')
             sage: lazy_foo[1] = 100
             sage: print(lazy_foo)
             [0, 100, 2, 3, 4, 5, 6, 7, 8, 9]
-            sage: sage.all.foo
+            sage: sage.all__sagemath_objects.foo
             [0, 100, 2, 3, 4, 5, 6, 7, 8, 9]
         """
         self.get_object()[key] = value
@@ -595,13 +595,13 @@ cdef class LazyImport():
         TESTS::
 
             sage: from sage.misc.lazy_import import LazyImport
-            sage: import sage.all
-            sage: sage.all.foo = list(range(10))
-            sage: lazy_foo = LazyImport('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = list(range(10))
+            sage: lazy_foo = LazyImport('sage.all__sagemath_objects', 'foo')
             sage: del lazy_foo[1]
             sage: print(lazy_foo)
             [0, 2, 3, 4, 5, 6, 7, 8, 9]
-            sage: print(sage.all.foo)
+            sage: print(sage.all__sagemath_objects.foo)
             [0, 2, 3, 4, 5, 6, 7, 8, 9]
         """
         del self.get_object()[key]
@@ -636,9 +636,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo + 1
             11
         """
@@ -648,9 +648,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo - 1
             9
         """
@@ -660,9 +660,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo * 2
             20
         """
@@ -674,9 +674,9 @@ cdef class LazyImport():
 
             sage: # needs sympy
             sage: from sympy import Matrix
-            sage: import sage.all
-            sage: sage.all.foo = Matrix([[1,1], [0,1]])
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = Matrix([[1,1], [0,1]])
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo.__matmul__(foo)
             Matrix([
             [1, 2],
@@ -688,9 +688,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo  // 3
             3
         """
@@ -700,9 +700,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: operator.truediv(foo, 3)
             10/3
         """
@@ -712,9 +712,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo ** 2
             100
         """
@@ -724,9 +724,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo % 7
             3
         """
@@ -736,9 +736,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo << 3
             80
         """
@@ -748,9 +748,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo >> 2
             2
         """
@@ -760,9 +760,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo & 7
             2
         """
@@ -772,9 +772,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo | 7
             15
         """
@@ -784,9 +784,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: foo ^^ 7
             13
         """
@@ -796,9 +796,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: -foo
             -10
         """
@@ -808,9 +808,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: +foo
             10
         """
@@ -820,9 +820,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = -1000
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = -1000
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: abs(foo)
             1000
         """
@@ -832,9 +832,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: ~foo
             1/10
         """
@@ -844,9 +844,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: complex(foo)
             (10+0j)
         """
@@ -856,9 +856,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: int(foo)
             10
         """
@@ -868,9 +868,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: float(foo)
             10.0
         """
@@ -880,9 +880,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: oct(foo)
             '0o12'
         """
@@ -892,9 +892,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: hex(foo)
             '0xa'
         """
@@ -904,9 +904,9 @@ cdef class LazyImport():
         """
         TESTS::
 
-            sage: import sage.all
-            sage: sage.all.foo = 10
-            sage: lazy_import('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = 10
+            sage: lazy_import('sage.all__sagemath_objects', 'foo')
             sage: list(range(100))[foo]
             10
         """
@@ -919,13 +919,13 @@ cdef class LazyImport():
         TESTS::
 
             sage: from sage.misc.lazy_import import LazyImport
-            sage: import sage.all
-            sage: sage.all.foo = [[1,2], 3]
-            sage: lazy_foo = LazyImport('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = [[1,2], 3]
+            sage: lazy_foo = LazyImport('sage.all__sagemath_objects', 'foo')
             sage: a = copy(lazy_foo)
-            sage: a is sage.all.foo        # copy
+            sage: a is sage.all__sagemath_objects.foo        # copy
             False
-            sage: a[0] is sage.all.foo[0]  # copy but not deep
+            sage: a[0] is sage.all__sagemath_objects.foo[0]  # copy but not deep
             True
             sage: type(lazy_foo) is LazyImport
             True
@@ -940,13 +940,13 @@ cdef class LazyImport():
         TESTS::
 
             sage: from sage.misc.lazy_import import LazyImport
-            sage: import sage.all
-            sage: sage.all.foo = [[1,2], 3]
-            sage: lazy_foo = LazyImport('sage.all', 'foo')
+            sage: import sage.all__sagemath_objects
+            sage: sage.all__sagemath_objects.foo = [[1,2], 3]
+            sage: lazy_foo = LazyImport('sage.all__sagemath_objects', 'foo')
             sage: a = deepcopy(lazy_foo)
-            sage: a is sage.all.foo        # copy
+            sage: a is sage.all__sagemath_objects.foo        # copy
             False
-            sage: a[0] is sage.all.foo[0]  # deep copy
+            sage: a[0] is sage.all__sagemath_objects.foo[0]  # deep copy
             False
             sage: type(lazy_foo) is LazyImport
             True
diff --git a/src/sage/misc/meson.build b/src/sage/misc/meson.build
index 8ceedd69ff5..5a6e139e39d 100644
--- a/src/sage/misc/meson.build
+++ b/src/sage/misc/meson.build
@@ -2,6 +2,9 @@ py.install_sources(
   '__init__.py',
   'abstract_method.py',
   'all.py',
+  'all__sagemath_environment.py',
+  'all__sagemath_objects.py',
+  'all__sagemath_repl.py',
   'allocator.pxd',
   'banner.py',
   'benchmark.py',
diff --git a/src/sage/misc/sageinspect.py b/src/sage/misc/sageinspect.py
index 9746bc34845..f2973d8f0f7 100644
--- a/src/sage/misc/sageinspect.py
+++ b/src/sage/misc/sageinspect.py
@@ -109,13 +109,13 @@ AUTHORS:
 """
 
 import ast
-import functools
 import inspect
+import functools
 import os
-import re
 import sys
 import tokenize
-from inspect import Parameter, Signature
+import re
+from inspect import Signature, Parameter
 
 try:
     import importlib.machinery as import_machinery
@@ -1336,11 +1336,12 @@ def sage_getfile(obj):
         if isinstance(obj, functools.partial):
             return sage_getfile(obj.func)
         return sage_getfile(obj.__class__)  # inspect.getabsfile(obj.__class__)
-    elif hasattr(obj, '__init__'):
-        pos = _extract_embedded_position(_sage_getdoc_unformatted(obj.__init__))
-        if pos is not None:
-            (_, filename, _) = pos
-            return filename
+    else:
+        if hasattr(obj, '__init__'):
+            pos = _extract_embedded_position(_sage_getdoc_unformatted(obj.__init__))
+            if pos is not None:
+                (_, filename, _) = pos
+                return filename
 
     # No go? fall back to inspect.
     try:
@@ -1387,7 +1388,7 @@ def sage_getfile_relative(obj):
     if not filename:
         return filename
 
-    from os.path import commonprefix, normpath, relpath
+    from os.path import relpath, normpath, commonprefix
 
     def directories():
         try:
@@ -1608,8 +1609,8 @@ def sage_getargspec(obj):
         sage: shell.run_cell('f??')
         ...the source code string...
     """
-    from sage.misc.abstract_method import AbstractMethod
     from sage.misc.lazy_attribute import lazy_attribute
+    from sage.misc.abstract_method import AbstractMethod
     if inspect.isclass(obj):
         return sage_getargspec(obj.__call__)
     if isinstance(obj, (lazy_attribute, AbstractMethod)):
@@ -1954,8 +1955,9 @@ def sage_formatargspec(args, varargs=None, varkw=None, defaults=None,
         specs.append(spec)
     if varargs is not None:
         specs.append(formatvarargs(formatargandannotation(varargs)))
-    elif kwonlyargs:
-        specs.append('*')
+    else:
+        if kwonlyargs:
+            specs.append('*')
     if kwonlyargs:
         for kwonlyarg in kwonlyargs:
             spec = formatargandannotation(kwonlyarg)
@@ -2663,6 +2665,10 @@ def __internal_tests():
         sage: sage_getdoc(None)
         ''
 
+        sage: import sage.all__sagemath_objects
+        sage: sage_getsource(sage.all__sagemath_objects)
+        '...all...'
+
     A cython function with default arguments (one of which is a string)::
 
         sage: sage_getdef(sage.rings.integer.Integer.binomial, obj_name='binomial')
diff --git a/src/sage/numerical/all__sagemath_polyhedra.py b/src/sage/numerical/all__sagemath_polyhedra.py
new file mode 100644
index 00000000000..6de2a9af502
--- /dev/null
+++ b/src/sage/numerical/all__sagemath_polyhedra.py
@@ -0,0 +1,9 @@
+from sage.misc.lazy_import import lazy_import
+
+lazy_import("sage.numerical.mip", ["MixedIntegerLinearProgram"])
+lazy_import("sage.numerical.sdp", ["SemidefiniteProgram"])
+lazy_import("sage.numerical.backends.generic_backend", ["default_mip_solver"])
+lazy_import("sage.numerical.backends.generic_sdp_backend", ["default_sdp_solver"])
+
+lazy_import("sage.numerical.interactive_simplex_method",
+            ["InteractiveLPProblem", "InteractiveLPProblemStandardForm"])
diff --git a/src/sage/numerical/backends/all__sagemath_polyhedra.py b/src/sage/numerical/backends/all__sagemath_polyhedra.py
new file mode 100644
index 00000000000..e69de29bb2d
diff --git a/src/sage/numerical/backends/meson.build b/src/sage/numerical/backends/meson.build
index 193ad419016..41c64b1af20 100644
--- a/src/sage/numerical/backends/meson.build
+++ b/src/sage/numerical/backends/meson.build
@@ -4,6 +4,7 @@ glpk = cc.find_library('glpk', required: not is_windows, disabler: true)
 py.install_sources(
   '__init__.py',
   'all.py',
+  'all__sagemath_polyhedra.py',
   'cvxopt_backend_test.py',
   'cvxpy_backend.pxd',
   'cvxpy_backend_test.py',
diff --git a/src/sage/numerical/meson.build b/src/sage/numerical/meson.build
index c14593d93ed..91257af0880 100644
--- a/src/sage/numerical/meson.build
+++ b/src/sage/numerical/meson.build
@@ -1,6 +1,7 @@
 py.install_sources(
   '__init__.py',
   'all.py',
+  'all__sagemath_polyhedra.py',
   'interactive_simplex_method.py',
   'knapsack.py',
   'linear_functions.pxd',
diff --git a/src/sage/repl/ipython_extension.py b/src/sage/repl/ipython_extension.py
index e5e1d6570ce..a4dd95648fe 100644
--- a/src/sage/repl/ipython_extension.py
+++ b/src/sage/repl/ipython_extension.py
@@ -617,7 +617,10 @@ class SageCustomizations:
             sage: SageCustomizations.all_globals()
             <module 'sage.all_cmdline' ...>
         """
-        from sage import all_cmdline
+        try:
+            from sage import all_cmdline
+        except ImportError:
+            from sage import all__sagemath_repl as all_cmdline
         return all_cmdline
 
     def init_environment(self):
diff --git a/src/sage_setup/autogen/interpreters/internal/__init__.py b/src/sage_setup/autogen/interpreters/internal/__init__.py
index 7d5090ac750..c2489ecd486 100644
--- a/src/sage_setup/autogen/interpreters/internal/__init__.py
+++ b/src/sage_setup/autogen/interpreters/internal/__init__.py
@@ -189,6 +189,17 @@ def rebuild(dirname, force=False, interpreters=None, distribution=None):
         sage: with open(testdir + '/wrapper_el.pyx') as f:
         ....:     f.readline()
         '# Automatically generated by ...\n'
+
+    Modularized build::
+
+        sage: testdir = tmp_dir()
+        sage: rebuild(testdir, interpreters=['Element', 'Python'],
+        ....:         distribution='sagemath-categories')
+        Generating interpreters for fast_callable in ...
+        -> First build of interpreters
+        sage: with open(testdir + '/all__sagemath_categories.py') as f:
+        ....:     f.readline()
+        '# Automatically generated by ...'
     """
     # This line will show up in "sage -b" (once per upgrade, not every time
     # you run it).
